### PR TITLE
Expose ruby_plugin on filter/output delegators

### PR DESCRIPTION
**Pull request by kaisecheng**
## What does this PR do?

Add a `ruby_plugin` JRuby method to all filter and output delegator types so that callers can access the underlying Ruby plugin instance when inspecting plugin config or instance variables.

Ruby-backed delegators (`FilterDelegatorExt`, `OutputDelegatorExt`, `OutputStrategyExt`) return the inner plugin object. Java-native delegators (`JavaFilterDelegatorExt`, `JavaOutputDelegatorExt`) return `nil` because they use the Java config system rather than `LogStash::Config::Mixin`.

## Why is it important/What is the impact to the user?

Filters and outputs are wrapped in Java delegators whose ancestor chain is the Java class hierarchy. Code that needs to inspect plugin config (e.g. reading SSL certificate paths for hot-reload) must unwrap the delegator first to access `get_config` and instance variables. This PR provides a standard unwrapping method across all delegator types.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~I have made corresponding changes to the documentation~~
- ~~I have made corresponding change to the default configuration files~~

## Related issues

- Part of https://github.com/elastic/logstash/issues/18892

### CODE REVIEW

File: logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java [hunk 1]
```diff
@@ -214,6 +214,11 @@ public IRubyObject workers() {
             return workers;
         }
 
+        @Override
+        public IRubyObject getRubyPlugin(final ThreadContext context) {
+            return workers.isEmpty() ? context.nil : workers.eltInternal(0);
```
Comment (andsel) [inline logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java]: Why using `eltInternal` instead of the `get(idx)` ?

File: logstash-core/src/test/java/org/logstash/config/ir/compiler/OutputDelegatorTest.java [hunk 2]
```diff
@@ -166,6 +166,11 @@ public void outputStrategyTests() {
 
             // test that metrics are properly set on the instance
             assertEquals(outputDelegator.namespacedMetric(), FakeOutClass.latestInstance.getMetricArgs());
+
+            // test that rubyPlugin returns the inner plugin instance
+            IRubyObject rubyPlugin = outputDelegator.rubyPlugin(RUBY.getCurrentContext());
+            assertThat(rubyPlugin).isInstanceOf(FakeOutClass.class);
+            assertThat(rubyPlugin.isNil()).isFalse();
```
Comment (andsel) [inline logstash-core/src/test/java/org/logstash/config/ir/compiler/OutputDelegatorTest.java:179]: This test method is becoming quite long, in its assertions, maybe this verification code could be extracted in its `verifyReturnInnerPluginInstance` method or similar.

File: logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java (hunk 1 above)
Comment (yaauie) [inline logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java]: If an output has this strategy, we will only return the plugin instance associated with the _first_ worker.

What are the ramifications here, or the opportunities for future "gotcha"s?

The stated purpose of this PR is to allow tooling to look at the instance's config (which should be the same across all instances) and to inspect ivars (which may _not_ be the same from one instance to the next).

Comment (kaisecheng) [reply logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java]: get() does more checking while eltInternal() is a direct access to the item. Both can be used.

Comment (kaisecheng) [reply logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java]: Can you help me to understand which plugin is using this Legacy class?

I couldn't find any plugin using it. I think it is leftover code that can potentially be removed. But I could be wrong. I pick the first one for backward compatibility. 

An alternative is to throw `UnsupportedOperationException`.

File: logstash-core/src/test/java/org/logstash/config/ir/compiler/OutputDelegatorTest.java (hunk 2 above)
Comment (kaisecheng) [reply logstash-core/src/test/java/org/logstash/config/ir/compiler/OutputDelegatorTest.java:179]: um... I think the test is not too bad that checks four states related to initiation. I feel that extracting the assertion adds an indirection layer. If you don't mind, I prefer to keep it.

**Comment by kaisecheng** [comment]
@andsel thanks for the review. I add a test for java null case.

File: logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java (hunk 1 above)
Comment (andsel) [reply logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java]: I've needed to check docs for `eltInternal` to know what it does. If there isn't appreciable performance penalty I would opt for the more explanatory `get`.

**Review by andsel — CHANGES_REQUESTED** [review]
I've left a preference on `get` method, for readability reason.
After that LGTM, but let's wait for @yaauie comment on https://github.com/elastic/logstash/pull/18936/changes#r3046417669 to understand better his concerns on picking the first instance of the list.

It's true that client code could access instance variables, and those differs instance by instance, by I think that in the context of your future changes it will bring just the configuration, which is the same for all instances. Let see if something pops up, maybe we can provide just the config part, directly from the plugin reference, instead of returning the reference to the complete plugin instance (config + variables).

File: logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java (hunk 1 above)
Comment (kaisecheng) [inline logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java]:
```suggestion
            return workers.isEmpty() ? context.nil : workers.get(0);
```

Comment (kaisecheng) [reply logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java]: changed to `get` for readability

File: logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java [hunk 3]
```diff
@@ -214,6 +214,11 @@ public IRubyObject workers() {
             return workers;
         }
 
+        @Override
+        public IRubyObject getRubyPlugin(final ThreadContext context) {
+            return workers.isEmpty() ? context.nil : workers.get(0);
```
Comment (kaisecheng) [inline logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java]:
```suggestion
            return workers.isEmpty() ? context.nil : (IRubyObject) workers.get(0);
```

File: logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java (hunk 1 above)
Comment (yaauie) [reply logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java]: Any output plugin that doesn't explicitly declare its concurrency [will default to `legacy`](https://github.com/elastic/logstash/blob/v5.0.0/logstash-core/lib/logstash/outputs/base.rb#L65-L67).

But it looks like the legacy strategy has been kneecapped since [at least v5.0.0](https://github.com/elastic/logstash/blob/v5.0.0/logstash-core/lib/logstash/outputs/base.rb#L65-L67), and doesn't allow anything other than a single "worker", which means the output strategy will only ever create one instance of the plugin.

It looks like my complaint is moot.

However, would it be prudent to add a method that lets callers _query_ the plugin and/or its config without returning the plugin instance itself? What _specifically_ are we needing to interrogate?

Comment (kaisecheng) [reply logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java]: ~Good to know that we can potentially remove the Legacy class.~

I am looking at this from a different angle. On the Ruby side, input plugins can be accessed directly without much friction, but the Java delegator currently prevents Ruby side (agent.rb) from accessing the underlying Ruby plugin object.

Now that we confirmed there is only a single inner plugin instance, returning that instance would bring output plugins in line with input plugins in terms of flexibility.

Unless you have additional concerns, I would prefer to keep this approach for future flexibility rather than introducing a more limited query interface.

Comment (kaisecheng) [reply logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java]: @andsel @yaauie Given that Legacy class only allows one worker and exposing the output plugin instance aligns with the behaviour of input plugins, can we move this PR forward? In case you spot additional flaw, I can create follow-up issue.

Comment (andsel) [reply logstash-core/src/main/java/org/logstash/config/ir/compiler/OutputStrategyExt.java]: I'm in favor of moving it forward.
